vhost-device-sound/src/stream.rs

```diff
@@ -239,85 +245,108 @@ impl Default for PcmParams {
     pub message: Arc<IOMessage>,
     direction: Direction,
 }
 
-impl std::fmt::Debug for Buffer {
+impl std::fmt::Debug for Request {
     fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
-        fmt.debug_struct(stringify!(Buffer))
+        fmt.debug_struct(stringify!(Request))
             .field("pos", &self.pos)
+            .field("len", &self.len)
             .field("direction", &self.direction)
             .field("message", &Arc::as_ptr(&self.message))
             .finish()
     }
 }
 
-impl Buffer {
-    pub fn new(
-        data_descriptor: virtio_queue::Descriptor,
-        message: Arc<IOMessage>,
-        direction: Direction,
-    ) -> Self {
+impl Request {
+    pub fn new(len: usize, message: Arc<IOMessage>, direction: Direction) -> Self {
         Self {
             pos: 0,
-            data_descriptor,
+            len,
             message,
             direction,
         }
     }
 
     pub fn read_output(&self, buf: &mut [u8]) -> Result<u32> {
-        let addr = self.data_descriptor.addr();
-        let offset = self.pos as u64;
-        let len = self
+        let mem = self.message.desc_chain.memory();
+
+        let mut reader = self
             .message
             .desc_chain
-            .memory()
-            .read(
-                buf,
-                addr.checked_add(offset)
-                    .expect("invalid guest memory address"),
-            )
+            .clone()
+            .reader(mem)
+            .map_err(|_| Error::DescriptorReadFailed)?;
+
+        let mut reader_content = reader
+            .split_at(size_of::<VirtioSoundPcmXfer>() + self.pos)
             .map_err(|_| Error::DescriptorReadFailed)?;
-        Ok(len as u32)
+
+        let bytes_read = reader_content
+            .read(buf)
+            .map_err(|_| Error::DescriptorReadFailed)?;
+
+        Ok(bytes_read as u32)
     }
 
     pub fn write_input(&mut self, buf: &[u8]) -> Result<u32> {
-        if self.desc_len() <= self.pos as u32 {
-            return Ok(0);
-        }
-        let addr = self.data_descriptor.addr();
-        let offset = self.pos as u64;
-        let len = self
+        let mem = self.message.desc_chain.memory();
+
+        let mut writer = self
             .message
             .desc_chain
-            .memory()
-            .write(
-                buf,
-                addr.checked_add(offset)
-                    .expect("invalid guest memory address"),
-            )
+            .clone()
+            .writer(mem)
+            .map_err(|_| Error::DescriptorReadFailed)?;
+
+        let mut _status = writer
+            .split_at(self.len)
+            .map_err(|_| Error::DescriptorReadFailed)?;
+
+        let mut write_content = writer
+            .split_at(self.pos)
+            .map_err(|_| Error::DescriptorReadFailed)?;
+
+        let bytes_written = write_content
+            .write(buf)
             .map_err(|_| Error::DescriptorWriteFailed)?;
-        self.pos += len;
```

Comment:
This change may be wrong.

Reply:
I reverted this change since it breaks alsa audio backend. This change did not break pipewire because we are incrementing in the audio backend too. I think the correct thing is to fix pipewire so we do not increment two times and we only increment during `write_input()`.